Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift

```diff
@@ -0,0 +1,231 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import struct Foundation.Date
+import class Foundation.JSONDecoder
+import struct Foundation.NSRange
+import class Foundation.NSRegularExpression
+import struct Foundation.URL
+import PackageModel
+import TSCBasic
+
+struct GitHubPackageMetadataProvider: PackageMetadataProvider {
+    let configuration: Configuration
+    let httpClient: HTTPClient
+    let decoder: JSONDecoder
+    let queue: DispatchQueue
+
+    init(configuration: Configuration = .init(), httpClient: HTTPClient = .init()) {
+        self.configuration = configuration
+        self.httpClient = httpClient
+        self.decoder = JSONDecoder()
+        #if os(Linux)
+        self.decoder.dateDecodingStrategy = .iso8601
+        #else
+        if #available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+            self.decoder.dateDecodingStrategy = .iso8601
+        } else {
+            self.decoder.dateDecodingStrategy = .customISO8601
+        }
+        #endif
+        self.queue = DispatchQueue(label: "org.swift.swiftpm.GitHubPackageMetadataProvider", attributes: .concurrent)
+    }
+
+    func get(_ reference: PackageReference, callback: @escaping (Result<PackageCollectionsModel.PackageBasicMetadata?, Error>) -> Void) {
+        guard reference.kind == .remote else {
+            return callback(.success(nil))
+        }
+        guard let baseURL = self.apiURL(reference.path) else {
+            return callback(.success(nil))
+        }
+
+        let metadataURL = baseURL
+        let tagsURL = baseURL.appendingPathComponent("tags")
+        let contributorsURL = baseURL.appendingPathComponent("contributors")
+        let readmeURL = baseURL.appendingPathComponent("readme")
+
+        self.queue.async {
+            // fan out multiple API calls
+
+            let urls = [metadataURL, tagsURL, contributorsURL, readmeURL]
+            var results = [URL: Result<HTTPClientResponse, Error>]()
+            let resultsLock = Lock()
+            let sync = DispatchGroup()
+            urls.forEach { url in
+                sync.enter()
+                let options = self.makeRequestOptions(validResponseCodes: [200])
+                httpClient.get(url, options: options) { result in
+                    defer { sync.leave() }
+                    resultsLock.withLock {
+                        results[url] = result
+                    }
+                }
+            }
+            sync.wait()
+
+            // process results
+
+            do {
+                let errors = results.compactMapValues { $0.failure }
+                // is this enough, or should other errors matter too?
```

Comment:
👀 @yim-lee

Reply:
The main error is repo not found (e.g., non-GH), so maybe we'd want to trap 404 and return NotFound error instead, otherwise this is enough IMO.